### ISSUE TITLE
Implement O_MAYEXEC / trusted_for()

### ISSUE THREAD

**Issue by kees** [KSPP/linux#164]
While executable files can be made non-executable, and mount points can be mounted with "noexec", there isn't a good way (short of comprehensive LSM policy) to block running an arbitrary shared library or script with an executable from such a place. This "noexec bypass using interpreters" is a long-standing hole in userspace defenses and there needs to be a way to check with the kernel on the origin of a given file to see if it should be allowed to be loaded by an interpreter.

Various versions of this functionality have existed in other distro patches for a while now (ClipOS and Chrome OS at least), but this needs to be done in upstream.

This series implements the trusted_for() syscall, which needs interpreters to use to validate permissions:
https://lore.kernel.org/lkml/20211008104840.1733385-1-mic@digikod.net

But we likely need to use the original design, based on Linus's feedback:
https://lore.kernel.org/lkml/20200723171227.446711-1-mic@digikod.net/

Basically only this patch remains, but needs to work with `faccessat` via `AT_EACCESS` instead of via `open`:
https://lore.kernel.org/lkml/20200723171227.446711-5-mic@digikod.net/
(and it must be strictly fd-based to avoid ToCToU on path-based APIs)

On the interpreter side, various things need to be considered (`-c`, stdin, etc). Some references:
- PEP 578 https://peps.python.org/pep-0578/
- CLIP OS patches https://github.com/search?q=repo%3Aclipos-archive%2Fclipos4_portage-overlay+o_mayexec&type=code

**Comment by l0kod** [comment]
Sounds good, I'm working on this.

**Comment by l0kod** [comment]
I sent a [v17](https://lore.kernel.org/r/20211115185304.198460-1-mic@digikod.net) to address a minor issue.

**Comment by kees** [comment]
[PR sent](https://lore.kernel.org/lkml/20220321161557.495388-1-mic@digikod.net/) but Linus wants a redesign.

**Comment by l0kod** [comment]
[[RFC PATCH v19 0/5] Script execution control (was O_MAYEXEC)](https://lore.kernel.org/all/20240704190137.696169-1-mic@digikod.net/)

**Comment by l0kod** [comment]
Hopefully latest patch series (now in -next): [[PATCH v23 0/8] Script execution control (was O_MAYEXEC)](https://lore.kernel.org/all/20241212174223.389435-1-mic@digikod.net/)

**Comment by l0kod** [comment]
[Merged in Linux 6.14](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=21266b8df5224c4f677acf9f353eecc9094731f0) 🥳
